File: src/main/java/net/clgd/ccemux/emulation/tror/TRoRTerminal.java
package net.clgd.ccemux.emulation.tror;

import dan200.computercraft.core.terminal.Terminal;
import java.util.LinkedList;
import net.clgd.ccemux.emulation.CCEmuX;

import static net.clgd.ccemux.Utils.BASE_16;

public class TRoRTerminal extends Terminal {
	public final LinkedList<TRoRPacket<? extends Object>> trorQueue = new LinkedList<>();

	public final CCEmuX emu;

	public TRoRTerminal(CCEmuX emu, int termWidth, int termHeight) {
		super(termWidth, termHeight);
		this.emu = emu;
	}

	protected void queue(TRoRPacket<? extends Object> packet) {
		synchronized (trorQueue) {
			trorQueue.add(packet);
		}
	}

	public LinkedList<TRoRPacket<? extends Object>> popQueue() {
		synchronized (trorQueue) {
			LinkedList<TRoRPacket<? extends Object>> out = new LinkedList<>(trorQueue);
			trorQueue.clear();
			return out;
		}
	}

	@Override
	public void resize(int w, int h) {
		super.resize(w, h);
		queue(new ResizePacket(w, h));
	}

	@Override
	public void setCursorPos(int x, int y) {
		super.setCursorPos(x, y);
		queue(new CursorPosPacket(x, y));
	}

	@Override
	public void setCursorBlink(boolean blink) {
		super.setCursorBlink(blink);
		queue(new CursorBlinkPacket(blink));
	}

	@Override
	public void setTextColour(int c) {
		super.setTextColour(c);
		queue(new TextColorPacket(BASE_16.charAt(c)));
	}

	@Override
	public void setBackgroundColour(int c) {
		super.setBackgroundColour(c);
		queue(new BackgroundColorPacket(BASE_16.charAt(c)));
	}

	@Override
	public void blit(String text, String fg, String bg) {
		super.blit(text, fg, bg);
		int y = getCursorY();
		queue(
			new BlitLinePacket(getLine(y).toString(), getTextColourLine(y).toString(), getBackgroundColourLine(y).toString()));
	}

	@Override
	public void write(String text) {
		super.write(text);
		queue(new WritePacket(text));
	}

	@Override
	public void scroll(int lines) {
		super.scroll(lines);
		queue(new ScrollPacket(lines));
	}

	@Override
	public void clear() {
		super.clear();
		queue(new ClearPacket());
	}

	@Override
	public void clearLine() {
		super.clearLine();
		queue(new ClearLinePacket());
	}

	@Override
	public void setLine(int line, String text, String fg, String bg) {
		super.setLine(line, text, fg, bg);
		queue(new CursorPosPacket(1, line));
		queue(new BlitLinePacket(text, fg, bg));
	}
}

File: src/main/kotlin/net/clgd/ccemux/emulation/tror/TRoRTerminal.kt
package net.clgd.ccemux.emulation.tror

import dan200.computercraft.core.terminal.Terminal
import net.clgd.ccemux.Utils.BASE_16
import net.clgd.ccemux.emulation.CCEmuX
import java.util.*

class TRoRTerminal(val emu: CCEmuX, termWidth: Int, termHeight: Int) : Terminal(termWidth, termHeight) {
	val trorQueue = LinkedList<TRoRPacket<*>>()

	protected fun queue(packet: TRoRPacket<*>) {
		synchronized (trorQueue) {
			trorQueue.add(packet)
		}
	}

	fun popQueue(): LinkedList<TRoRPacket<*>> {
		synchronized (trorQueue) {
			@Suppress("UNCHECKED_CAST") // no other way around this :/
			val out = trorQueue.clone() as LinkedList<TRoRPacket<*>>
			trorQueue.clear()
			return out
		}
	}

	override fun resize(w: Int, h: Int) {
		super.resize(w, h)
		queue(ResizePacket(w, h))
	}

	override fun setCursorPos(x: Int, y: Int) {
		super.setCursorPos(x, y)
		queue(CursorPosPacket(x, y))
	}

	override fun setCursorBlink(blink: Boolean) {
		super.setCursorBlink(blink)
		queue(CursorBlinkPacket(blink))
	}

	override fun setTextColour(c: Int) {
		super.setTextColour(c)
		queue(TextColourPacket(BASE_16.getOrElse(c, { '0' })))
	}

	override fun setBackgroundColour(c: Int) {
		super.setBackgroundColour(c)
		queue(BackgroundColourPacket(BASE_16.getOrElse(c, { 'f' })))
	}

	override fun blit(text: String, fg: String, bg: String) {
		super.blit(text, fg, bg)
		val y = cursorY
		queue(BlitLinePacket(getLine(y).toString(), getTextColourLine(y).toString(), getBackgroundColourLine(y).toString()))
	}

	override fun write(text: String) {
		super.write(text)
		queue(WritePacket(text))
	}

	override fun scroll(lines: Int) {
		super.scroll(lines)
		queue(ScrollPacket(lines))
	}

	override fun clear() {
		super.clear()
		queue(ClearPacket())
	}

	override fun clearLine() {
		super.clearLine()
		queue(ClearLinePacket())
	}

	override fun setLine(line: Int, text: String, fg: String, bg: String) {
		super.setLine(line, text, fg, bg)
		queue(CursorPosPacket(1, line))
		queue(BlitLinePacket(text, fg, bg))
	}
}
